freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/handler/UpgradeCcmUpgradeHandler.java

```diff
@@ -35,26 +42,30 @@ public String selector() {
     @Override
     protected Selectable defaultFailureEvent(Long resourceId, Exception e, Event<UpgradeCcmEvent> event) {
         LOGGER.error("Running upgrade for CCM upgrade has failed", e);
-        return new UpgradeCcmFailureEvent(UPGRADE_CCM_FAILED_EVENT.event(), resourceId, event.getData().getOldTunnel(), getClass(), e);
+        LocalDateTime revertDate = LocalDateTime.now().plusMinutes(ccmParametersConfigService.getActivationInMinutes());
+        return new UpgradeCcmFailureEvent(UPGRADE_CCM_FAILED_REVERT_EVENT.event(), resourceId, event.getData().getOldTunnel(),
+                getClass(), e, revertDate, "Upgrading CCM is failed.");
     }
 
     @Override
     protected Selectable doAccept(HandlerEvent<UpgradeCcmEvent> event) {
         UpgradeCcmEvent request = event.getData();
-        upgradeCcmService.changeTunnel(request.getResourceId());
+        upgradeCcmService.changeTunnel(request.getResourceId(), Tunnel.latestUpgradeTarget());
+        UpgradeCcmEvent resultEvent = UPGRADE_CCM_UPGRADE_FINISHED_EVENT.createBasedOn(request);
         if (request.getOldTunnel().useCcmV1()) {
             try {
                 LOGGER.info("Running upgrade state for CCM...");
                 upgradeCcmService.upgrade(request.getResourceId());
+                resultEvent.setUpgradeDateTime(LocalDateTime.now().plusMinutes(ccmParametersConfigService.getActivationInMinutes()));
             } catch (CloudbreakOrchestratorException e) {
                 LOGGER.debug("Failed applying CCM upgrade state");
-                return new UpgradeCcmFailureEvent(UPGRADE_CCM_FAILED_EVENT.event(), request.getResourceId(),
-                        event.getData().getOldTunnel(), getClass(), e);
+                return new UpgradeCcmFailureEvent(UPGRADE_CCM_FAILED_REVERT_EVENT.event(), request.getResourceId(),
+                        event.getData().getOldTunnel(), getClass(), e, event.getData().getRevertTime(), "Upgrading CCM is failed.");
             }
         } else {
             LOGGER.info("Running upgrade step is skipped for previous tunnel type '{}'", request.getOldTunnel());
         }
-        return UPGRADE_CCM_UPGRADE_FINISHED_EVENT.createBasedOn(request);
+        return resultEvent;
     }
```

Comment:
If you refactor this a bit it's more readable imho. also please add a log at least where I added a placeholder
```
    protected Selectable doAccept(HandlerEvent<UpgradeCcmEvent> event) {
        UpgradeCcmEvent request = event.getData();
        upgradeCcmService.changeTunnel(request.getResourceId(), Tunnel.latestUpgradeTarget());
        if (request.getOldTunnel().useCcmV1()) {
            return tryToUpgradeCccm(event, request);
        } else {
            LOGGER.info("Running upgrade step is skipped for previous tunnel type '{}'", request.getOldTunnel());
            return UPGRADE_CCM_UPGRADE_FINISHED_EVENT.createBasedOn(request);
        }
    }

    private StackEvent tryToUpgradeCccm(HandlerEvent<UpgradeCcmEvent> event, UpgradeCcmEvent request) {
        try {
            LOGGER.info("Running upgrade state for CCM...");
            upgradeCcmService.upgrade(request.getResourceId());
            <LOG something>
            UpgradeCcmEvent resultEvent = UPGRADE_CCM_UPGRADE_FINISHED_EVENT.createBasedOn(request);
            resultEvent.setUpgradeDateTime(LocalDateTime.now().plusMinutes(ccmParametersConfigService.getActivationInMinutes()));
            return resultEvent;
        } catch (CloudbreakOrchestratorException e) {
            LOGGER.debug("Failed applying CCM upgrade state");
            return new UpgradeCcmFailureEvent(UPGRADE_CCM_FAILED_REVERT_EVENT.event(), request.getResourceId(),
                    event.getData().getOldTunnel(), getClass(), e, event.getData().getRevertTime(), "Upgrading CCM is failed.");
        }
    }

Reply:
ok